### PR TITLE
#1120 Improvements for Reviewer stats updates

### LINKED ISSUE #1120: Major bug: Assigning/Unassigning taking unreasonably long time
What we know:
- it's connected to the "update_application_reviewer_stats" function, which is triggered on review assignment updates (if I disable this trigger, it happens in the normal, timely manner)

What we don't know
- why only on the Live server (it's fast on Demo3 with the exact same snapshot, and that's a slower system)?
- why so slow? Have checked locally, and the number of database operations caused by the trigger shouldn't take this long, even on a slower system.

### Possible solution

In order to prevent this function blocking the update, we could re-implement it on the Server (rather than as a Postgres function), and trigger it from a notification channel. And then run is `async`, so it doesn't block the main mutation.

Still doesn't really address why it's causing such a block, but we'll be able to add more logging to it this way and have a better idea of what's causing it to run so many times.

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fix #1120 

### Main changes:
- Remove the "Insert" trigger for `review_assignment` and `review_assignment_assigner_join`, as this was causing the update function to run way too many times[^1] whenever review assignments were generated. Instead it just gets run once and is called by the `generateReviewAssignments` Action.
- Move the "updateReviewerStats" function to a server function, triggered by database notification (and other things) rather than a PG function. Makes it much easier to see when it's running
- Made sure "updateReviewerStats" gets called async and send to the Throttle queue, which means it has to wait until the current action sequence is complete, and won't block the UI (as the Action triggers don't wait for it)

### Additional related changes
- Make `updateReviewerStats` callable by command line, so we can run manually on existing applications if desired
- Make Throttle a global object so can be used by more than just the database trigger functions (and update types accordingly)
- Data migration to re-calculate reviewer actions for existing applications (that have reviewer actions currently), as we had a few where they were showing up as "Assignable" even though they were approved.

[^1]: It was running for a single reviewer for each insert, whereas now it runs once for all reviewers. I didn't realise quite how bad it was, due to two things:
    - even though it only runs for a single reviewer, the reviewer and assigner lists for the application table were being re-calculated every time, which was very wasteful
    - the `review_assignment_assigner_join` captures a many-to-many relation, so by calling the function on every insert, the number of potential assigners for each application can be quite large (they get repeated for *each* review assignment, so there was 112 in the case that was making this problem noticeable on the Fiji server). And because a lot of them were self-assigning, it's the same person for a lot of them. So we would be getting well over a hundred calls for what boiled down to about 10 actual users. i.e. if it had only been running once per person, it might have been okay, but once per assigner-assignment join was too much!

### CODE REVIEW

File: plugins/action_generate_review_assignment_records/src/generateReviewAssignments.test.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@
 // database has entered correct information
 
 import { map, omit } from 'lodash'
-import DBConnect from '../../../src/components/databaseConnect'
+import DBConnect from '../../../src/components/database/databaseConnect'
```
Comment (CarlosNZ) [inline plugins/action_generate_review_assignment_records/src/generateReviewAssignments.test.ts:7]: Should have been part of previous PR (#1122), but didn't notice until I tried to build and got errors

File: src/components/actions/throttle.ts [hunk 2]
```diff
@@ -21,8 +21,8 @@ const THROTTLE_QUEUE_THRESHOLD = 200 // ms
 
 type EventObject<T, U> = { name: string; data: T; action: (data: T) => Promise<U> | U }
 
-export class EventThrottle<T, U> {
-  queue: EventObject<T, U>[]
+export class EventThrottle {
+  queue: EventObject<any, any>[]
```
Comment (CarlosNZ) [inline src/components/actions/throttle.ts:25]: Okay, this has to be a truly legit use of `any`. I wanted to keep the Generic on the "EventObject", to enforce that the input "data" has to be consistent with the "action" function, but I *don't* want to have to explicitly specify a type when creating the "Event Throttle" object, as it should handle any type of Event object.

But, annoyingly, it requires a type in this definition, so it has to be `any`, right?

File: src/components/database/updateReviewerStats.ts [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+      break
+    case 'application':
+      applicationId = data.id as number
+      // Update ALL staff by leaving reviewerIds empty
+      break
+  }
+  updateReviewerStats(applicationId, reviewerIds.length === 0 ? undefined : reviewerIds)
+}
+
+const databaseMethods = {
```
Comment (CarlosNZ) [inline src/components/database/updateReviewerStats.ts:85]: SQL methods here all basically copied as-is from the previous PG function.

File: src/types.ts [hunk 4]
```diff
@@ -383,6 +384,7 @@ interface ConfigBase {
   isLiveServer: boolean
   emailMode: EmailOperationMode
   maintenanceMode: boolean
+  Throttle: EventThrottle
```
Comment (CarlosNZ) [inline src/types.ts:387]: The Throttle used to be only used in one place, but now we need to make the single instance globally available, hence part of config object.

File: src/components/actions/throttle.ts [hunk 5]
```diff
@@ -19,10 +19,14 @@
...
+interface EventObject<T> {
+  name: string
+  data: T
+  action: (data: T) => unknown
+}
 
-export class EventThrottle<T, U> {
-  queue: EventObject<T, U>[]
+export class EventThrottle {
+  queue: EventObject<any>[]
```
Comment (CarlosNZ) [inline src/components/actions/throttle.ts:29]: Okay, this has to be a truly legit use of `any`. I wanted to keep the Generic on the "EventObject", to enforce that the input "data" has to be consistent with the "action" function, but I *don't* want to have to explicitly specify a type when creating the "Event Throttle" object, as it should handle any type of Event object.

But, annoyingly, it requires a type in this definition, so it has to be `any`, right?

See [this](https://www.typescriptlang.org/play/?jsx=0#code/PTAEHUFNQCwQwG7TgO1AewEYCtIGMAXUAgTwAdoB3ASwJmMvVDICd0KWDrIBnAGgBQIUKgAmoSlVRECTSCgBm6Fnmh04RTOjrM2HLrxEto6BcRjQecALZrykQcMwBXIpNCj0KAORE8cYwl4GXs+UB4md2ceZFAAcXlIFmo8ADpQAAVjAlJwgjEA8R4CZwUFVIFqFAIkhThVUAARDTgAeRx8AgAeABUAPlAAbwFQXXYk0gBGAC5QHpGx-RIAJln5gF8BITAAIVdzeQloeCQRFG0LFgwOwmJQoJT6fzQYdAAbcVQjFjgSDDM6JBtk0Wu1cLcsOCCDx0lAPF5fBJpMQmMYAI7OaiBdQyCwg-Jgzo8UCYZBvN7-A7Aqy2O4UMIRWDoSgSdDOD5HeHmajEmh0NlEABEqBIgoA-JVqrV6tAAMJefJVJKE27DUYoGyQWbFZIoADmC1ELQAkjVrDxZs0CTduiK+gBtAC6AnWQA) for a simplified example of the case.

File: src/components/database/updateReviewerStats.ts [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+      const { application_id, reviewer_id, assigner_id } = await db.getApplicationDataFromReview(
+        data.review_id as number
+      )
+      applicationId = application_id
+      reviewerIds.push(...new Set([reviewer_id, assigner_id]))
+      break
+    case 'application':
+      applicationId = data.id as number
+      // Update ALL staff by leaving reviewerIds empty
+      break
```
Comment (andreievg) [inline src/components/database/updateReviewerStats.ts:80]: default case would make applicationId = 0 ?

File: src/components/actions/throttle.ts (hunk 5 above)
Comment (andreievg) [reply src/components/actions/throttle.ts:29]: Could do something like this: https://github.com/msupply-foundation/conforma-server/compare/1120-fix-trigger-block-for-review-stats-update...1120-fix-trigger-block-for-review-stats-update(No-any-example)

**Comment by andreievg** [comment]
I think it's worth handling an error in action queue: https://github.com/msupply-foundation/conforma-server/compare/1120-fix-trigger-block-for-review-stats-update...1120-fix-trigger-block-for-review-stats-update(Error-handling)

File: src/components/database/updateReviewerStats.ts [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+                SELECT id FROM public.review_assignment
+                WHERE application_id = $1
+            );
+        `
+      const result = await DBConnect.query({
+        text,
+        values: [appId],
+        rowMode: 'array',
+      })
+      return result.rows.flat()
```
Comment (andreievg) [inline src/components/database/updateReviewerStats.ts:124]: it's possible that reviewers will also be assigners ? I think we would want to dedup

Comment (CarlosNZ) [reply src/components/database/updateReviewerStats.ts:124]: The `DISTINCT` in each subquery and the `UNION` results in a list of unique IDs, no dupes (assigners and reviewers combined). You would think the `UNION` would combine them, but it doesn't in this case 🤷‍♂️

Comment (andreievg) [reply src/components/database/updateReviewerStats.ts:124]: Ohh true, looks like union does distict on rows https://www.postgresql.org/docs/current/queries-union.html

**Review by andreievg — APPROVED** [review]
Looks great thanks, I had a quick look and it seems pretty solid to me. I do have a slight concern about, trying to do granular updates, like when [review_status_history](https://github.com/msupply-foundation/conforma-server/blob/a3afd1ed21ef124d2b231b4f1fa135d740ecfb9d/src/components/database/updateReviewerStats.ts#L70-L74) has changed, we only look at reviewer and assigner, I think it would be simpler to just do updates for all of the application staff, but that could lead to excessive updates, especially if application is changed soon after. In which case I think i would create a new queue in database, with just application_ids and auto increment, and do a simple polling (every x seconds) to process it (grouping application_ids updates) <- I don't think it will come to that as the current logic looks pretty solid, thanks !

Is there a change that updateReviewStatsFromDBEvent will run multiple times for the same DB event ? (like when review_status_history update or review_assignment or application will be updated multiple times in sequence ?)

File: src/components/actions/throttle.ts (hunk 5 above)
Comment (CarlosNZ) [reply src/components/actions/throttle.ts:29]: Yeah, except I don't really like having to wrap the `action` into an additional anonymous function call just to keep Typescript happy. I like being able to pass the `action` and `data` in separately and calling `action` on `data` directly, and using the Generic to enforce consistency between action method and its input.

Minor thing, slightly prefer my way. (Plus, any thoughts about the even more general case in the [playground example](https://www.typescriptlang.org/play/?jsx=0#code/PTAEHUFNQCwQwG7TgO1AewEYCtIGMAXUAgTwAdoB3ASwJmMvVDICd0KWDrIBnAGgBQIUKgAmoSlVRECTSCgBm6Fnmh04RTOjrM2HLrxEto6BcRjQecALZrykQcMwBXIpNCj0KAORE8cYwl4GXs+UB4md2ceZFAAcXlIFmo8ADpQAAVjAlJwgjEA8R4CZwUFVIFqFAIkhThVUAARDTgAeRx8AgAeABUAPlAAbwFQXXYk0gBGAC5QHpGx-RIAJln5gF8BITAAIVdzeQloeCQRFG0LFgwOwmJQoJT6fzQYdAAbcVQjFjgSDDM6JBtk0Wu1cLcsOCCDx0lAPF5fBJpMQmMYAI7OaiBdQyCwg-Jgzo8UCYZBvN7-A7Aqy2O4UMIRWDoSgSdDOD5HeHmajEmh0NlEABEqBIgoA-JVqrV6tAAMJefJVJKE27DUYoGyQWbFZIoADmC1ELQAkjVrDxZs0CTduiK+gBtAC6AnWQA)?)

File: src/components/database/updateReviewerStats.ts (hunk 6 above)
Comment (CarlosNZ) [reply src/components/database/updateReviewerStats.ts:80]: Yeah, which would do nothing (tested the sql, doesn't throw any errors, but updates no records)

**Comment by CarlosNZ** [comment]
> I think it's worth handling an error in action queue: [1120-fix-trigger-block-for-review-stats-update...1120-fix-trigger-block-for-review-stats-update(Error-handling)](https://github.com/msupply-foundation/conforma-server/compare/1120-fix-trigger-block-for-review-stats-update...1120-fix-trigger-block-for-review-stats-update(Error-handling))

Yeah, good call: 833703de

**Comment by CarlosNZ** [comment]
> Is there a change that updateReviewStatsFromDBEvent will run multiple times for the same DB event ? (like when review_status_history update or review_assignment or application will be updated multiple times in sequence ?)

Not sure why that would happen. Previously there was a bit of a problem with the trigger field being set/updated/reset in quick succession causing multiple executions, hence I've tightened up those triggers (on `application` and `review_assignment`) to only respond to particular columns (i.e. NOT the trigger column).

And `review_status_history` only ever gets inserts, never updates (and only when the status changes).

**Comment by CarlosNZ** [comment]
Thanks for the review @andreievg 👍🏼

**Pull request merged**